File: content/chapters/compute/lab/support/sum-array/c/sum_array_sequential.c
// SPDX-License-Identifier: BSD-3-Clause

#include <stdlib.h>
#include <stdio.h>
#include <sys/time.h>
#include <errno.h>
#include <string.h>

#include "include/generate_random_array.h"
#include "include/array_utils.h"
#include "utils/utils.h"

int main(void)
{
	long result = 0;
	struct timeval start, stop;
	long time;
	int *array;

	array = malloc(sizeof(*array) * ARR_LEN);
	DIE(!array, "Error when allocating array");

	generate_random_array(ARR_LEN, array);

	gettimeofday(&start, NULL);

	for (int i = 0; i < ARR_LEN; i++)
		result += array[i];

	gettimeofday(&stop, NULL);

	time = 1000 * (stop.tv_sec - start.tv_sec) + (stop.tv_usec - start.tv_usec) / 1000;

	printf("Array sum is %ld\nTime spent: %lu miliseconds\n", result, time);

	free(array);

	return 0;
}

File: content/compute/lab/support/sum-array/python/sum_array_sequential.py
# SPDX-License-Identifier: BSD-3-Clause

from random import randint
from time import time_ns

ARR_LEN = 10_000_000


def generate_random_array(length=ARR_LEN):
    return [randint(1, 100) for _ in range(length)]


def main():
    arr = generate_random_array()
    sum_arr = 0
    time_start = time_ns()

    for elem in arr:
        sum_arr += elem

    time_end = time_ns()
    time_spent = (time_end - time_start) / 1000_000

    print(f"Array sum is: {sum_arr}; time spent: {time_spent}ms")


if __name__ == "__main__":
    exit(main())
